### ISSUE TITLE
Feature Request: support for Gradle

### ISSUE THREAD

**Issue by amirabiri** [mongock/mongock#386]
I tried to build a simple Gradle plugin for Mongock. I generally managed to get it to work - but only by adding a small change to Mongock.

I used a simple approach based on StandaloneRunner of course:

```kotlin
abstract class MongockMigrateTask

    // ...

    @TaskAction
    fun migrate() {
        val mongoClient = MongoClients.create(mongoConnectionString.get())
        val runner = MongockStandalone.builder()
            .setDriver(MongoSync4Driver.withDefaultLock(mongoClient, "dbName"))
            .addChangeLogsScanPackage(migrationsNamespace.get())
            .buildRunner()
        runner.execute()
    }
}
```

Out of the box this method doesn't work - The context in which the Gradle task runs does not contain the project's classes on the classpath.

Initially I tried a JavaExec based approach, but that's a huge nightmare.

So I checked how Flyway does it - and I noticed that [Flyway at the core allows configuring with a custom ClassLoader](https://github.com/flyway/flyway/blob/flyway-7.15.0/flyway-gradle-plugin/src/main/java/org/flywaydb/gradle/task/AbstractFlywayTask.java#L680).

So I checked out the code and tried it - and it works. [Here is my change](https://github.com/amirabiri/mongock-core/commit/63148ce4d90066963016a032fe2dfc48bff2b522).

It's a trivial change really, but I didn't do a PR because I didn't add any tests (wasn't sure how to test with custom class loader).

The final Gradle task required a bit more trickery but it wasn't hard:

```kotlin
abstract class MongockMigrateTask

    // ...

    @TaskAction
    fun migrate() {

        val extraURLs = HashSet<URL>()
        addClassesAndResourcesDirs(extraURLs)
        addConfigurationArtifacts(arrayOf("compileClasspath", "runtimeClasspath"), extraURLs)
        val classLoader = URLClassLoader(
            extraURLs.toTypedArray(),
            project.buildscript.classLoader
        )

        val mongoClient = MongoClients.create(mongoConnectionString.get())
        val runner = MongockStandalone.builder()
            .setDriver(MongoSync4Driver.withDefaultLock(mongoClient, "dbName"))
            .addChangeLogsScanPackage(migrationsNamespace.get())
            .setClassLoader(classLoader)
            .buildRunner()
        runner.execute()
    }

    private fun addClassesAndResourcesDirs(extraURLs: MutableSet<URL>) {
        // Lifted from Flyway...
    }

    private fun addConfigurationArtifacts(configurations: Array<String>, urls: MutableSet<URL>) {
        // Lifted from Flyway...
    }
}
```

Anyhow it proves that  it can be done and opens the door for more build systems.

----

**However** - I did come across one problem which raises an interesting question -

Since my project is Spring based, I used `MongockTemplate` - and my Gradle task threw a `ClassNotFound` exception. This is due to the fact that In standalone mode it can't inject `MongockTemplate` which is based on Spring's `MongoTemplate`.

That forced me to change the migration to use `MongoDatabase`, which is a shame:

```kotlin
@ChangeLog(order = "001")
class V001_InitCollection {

    @ChangeSet(order = "001", id = "init_collection", author = "me")
    fun initCollection(db: MongockTemplate) {
        if (!db.collectionExists("collection")) { // 😇
            db.createCollection("collection")
        }
    }
}
```

```kotlin
@ChangeLog(order = "001")
class V001_InitCollection {

    @ChangeSet(order = "001", id = "init_collection", author = "me")
    fun initCollection(db: MongoDatabase) {
        if (!db.listCollectionNames().contains("collection")) { // 🤮
            db.createCollection("collection")
        }
    }
}
```

I think it would be great to make `MongockTemplate` an interface that works exactly the same in all drivers.

**Comment by amirabiri** [comment]
I'm not sure I understand how this works... I have to compile a special CLI specific to my project which is linked to its sources?

If so this isn't what I was aiming for. I was hoping to go with the same workflow we have today with Flyway - i.e given a .jar with Flyway migrations (based on convention) - I can run the migrations using `flyway` cli, Maven, Gradle, or trigger from code with a simple `main()` ("standalone") - with no difference between them.

It is also framework agnostic. There is no need to compile some cli into my jar which isn't desirable because I do a lot of things with the CLI and I can't let a library take control over it.

In the end I use two methods overall in my workflow:
 - During development - run the migrations from Gradle.
 - In production - trigger from code using our own cli verb.

**Comment by dieppa** [comment]
Hello @amirabiri ,

I first going to clarify how this CLI works. 

To start it's in BETA version, we are still adjusting ways of working 😄 

Second, you are not adding the CLI to your jar(well, you bring the dependency,as you are doing with the runner), but what it's actually doing a new .jar from your code and just changing the start-class. So you .jar remains untouch, we create a new one, with a new manifest to use another starting class. I don't know if that changes your perspective.

Third, the reasons I do that is actually to load the entire context. With this approach, you have all the injections your want: MongoTemplate, your repositories, etc. 

I understand you still are not comfortable with this approach. It can be done also having a separate CLI to which you pass your .jar,  but if you want to have the flexibility of having the Spring conext loaded(or any other framework) in a robust manner, without falling into a headache, in reallity it won't change much, to be honest. Just you don't build the CLI specificcally to your project, you have a unique one instead, but a similar job needs to be done in another stage.

There is also the option to use the classLoader, but then you don't have the Spring Context(or any other)..at least no in a way that is not prone to errors.


**In summary, we can provide more ways to do the CLI and I would like to understad really your needs and how best solve them. What about having a zoom/team session about it??**

**Comment by dieppa** [comment]
Hello @amirabiri ,

just to let you know we are working on the CLI to allow the user to pass the application jar...and also solving the issue with MongockTemplate and any other dependency you have. 

We haven't finalised yet(maybe at the end of the week. Probably earlier), but it will look something like this(no maven plugin required):
`./mongock migrate -jar your_jar_path.jar `

**Does this work for you?**

**Comment by amirabiri** [comment]
Hi @dieppa,

Wow that's a very quick listening to your users! Thank you :)

I will check it out as soon as I can. 

What technique are you going with to support this? It should be possible to use it as well for the Gradle plugin.

**Comment by dieppa** [comment]
Hi @amirabiri ,

thanks for appreciating!! 😃 

the user will basically need yo run a .sh passing the project's jar and the operation and optional parameters. I think that fits well which what you need. Correct me if I am wrong.

In the near future we'll add more operations(we already have many in mind), for which you don't need to pass the jar, for example to know the state of the database, run ad hoc changes or connect to the dashboard(we'll be adding a dashboard too).

**Comment by dieppa** [comment]
Hello @amirabiri ,

have you been able to check this out?

**Comment by amirabiri** [comment]
Hi @dieppa,

I haven't been able to get it to work, I get an error:

```
./mongock --appJar /path/to/my/spring/boot/app
Exception in thread "main" java.lang.RuntimeException: Missing jar parameter. Please follow the format "--appJar jar_path"
	at io.mongock.cli.wrapper.MongockCliMain.getParameter(MongockCliMain.java:80)
	at io.mongock.cli.wrapper.MongockCliMain.main(MongockCliMain.java:31)
```

Regardless, this isn't what I was looking for. I would like to run the migrations from within Gradle like so:

```
./gradlew mongockMigrate
```

I guess this can be accomplished by gradle running the `mongock` cli, but sounds cumbersome and error prone to me. Have you looked at the diff I sent? https://github.com/amirabiri/mongock-core/commit/63148ce4d90066963016a032fe2dfc48bff2b522

All that is needed really is just to allow passing a custom ClassLoader to the runner.

**Comment by dieppa** [comment]
Can you update on the state of this ticket? 😄
